### PR TITLE
*: Change the engine behind `ContentPathReloader` to be completely independent of any filesystem concept.

### PR DESCRIPTION
**Pull request by douglascamata**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This should solve #6491.

* Replaces the config reload logic behind the `PathContentReloader` function, which now will frequently reread a file and run the `reloadFunc` whenever the sha256 of the file's contents changes.
* Updated the `PathContentReloader` tests accordingly, including new tests covering the new engine. 
* Adds a new CLI flag to Receive, `receive.limits-config-reload-timer`, used to configure the debouncer timer of the fs notification engine or the reload time of the polling engine.
  * By default this is set to `1s`, which matches the current behavior. There's no breaking change there.

## Verification

Automated tests. Ran time quite a few times to try to capture any flakiness.

### CODE REVIEW

**Comment by douglascamata** [comment]
Waiting on the results of a conversation that might culminate in removing the fsnotify engine and staying only with polling one.

File: pkg/extkingpin/path_content_reloader.go [hunk 1]
```diff
@@ -22,74 +20,65 @@ type fileContent interface {
...
+type pollingEngine struct {
+	filePath         string
+	logger           log.Logger
+	debounce         time.Duration
+	reloadFunc       func()
+	previousChecksum [sha256.Size]byte
+}
+
+func (p *pollingEngine) Start(ctx context.Context) error {
+	configReader := func() {
```
Comment (saswatamcode) [inline pkg/extkingpin/path_content_reloader.go:53]: We can probably have this as a separate private method defined on `pollingEngine`, and just have the goroutine defined in `Start` that calls it?

File: pkg/extkingpin/path_content_reloader.go [hunk 2]
```diff
@@ -22,74 +20,65 @@ type fileContent interface {
...
+			return
+		}
+		checksum := sha256.Sum256(file)
+		if checksum == p.previousChecksum {
+			return
 		}
-		defer watcher.Close()
+		p.reloadFunc()
+		p.previousChecksum = checksum
+		level.Debug(p.logger).Log("msg", "configuration reloaded")
```
Comment (saswatamcode) [inline pkg/extkingpin/path_content_reloader.go:70]: Let's add the filepath to the debug log too!
```suggestion
		level.Debug(p.logger).Log("msg", "configuration reloaded", "path", p.filePath)
```

File: pkg/extkingpin/path_content_reloader.go [hunk 3]
```diff
@@ -22,74 +20,65 @@ type fileContent interface {
...
-			reloadTimer = time.AfterFunc(debounceTime, func() {
-				reloadFunc()
-				level.Debug(logger).Log("msg", "configuration reloaded after debouncing")
-			})
-			reloadTimer.Stop()
+	return engine.Start(ctx)
+}
+
+// pollingEngine is an implementation of reloaderEngine that keeps rereading the contents at filePath and when the
+// checksum changes it runs the reloadFunc.
```
Comment (saswatamcode) [inline pkg/extkingpin/path_content_reloader.go:43]: No interface needed now 🙂 
```suggestion
// pollingEngine keeps rereading the contents at filePath and runs 
// reloadFunc when the checksum changes.
```

File: pkg/extkingpin/path_content_reloader.go [hunk 4]
```diff
@@ -22,74 +20,65 @@ type fileContent interface {
...
+	logger           log.Logger
+	debounce         time.Duration
+	reloadFunc       func()
+	previousChecksum [sha256.Size]byte
+}
+
+func (p *pollingEngine) Start(ctx context.Context) error {
+	configReader := func() {
+		// check if file still exists
+		if _, err := os.Stat(p.filePath); os.IsNotExist(err) {
```
Comment (saswatamcode) [inline pkg/extkingpin/path_content_reloader.go:55]: Do we need a separate `Stat` call, as we aren't doing anything different on the error?

**Review by saswatamcode — APPROVED** [review]
Thanks! LGTM mod some small nits!

**Review by saswatamcode — APPROVED** [review]
LGTM!

**Pull request merged**